### PR TITLE
Refactor the loadbalancer to match Java implementation

### PR DESCRIPTION
**Pull request by stevegury**
This PR refactor the loadbalancer to match the Java implementation.
The LB is now estimating latency and distributed load based on those prediction.

### CODE REVIEW

File: lib/common/frugalMedian.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+function FrugalMedian() {
```
Comment (yunong) [inline lib/common/frugalMedian.js]: No JS Docs for this?

File: lib/connection/drainingreactivesocket.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+function DrainingReactiveSocket(reactivesocket) {
```
Comment (yunong) [inline lib/connection/drainingreactivesocket.js]: JS Docs please.

File: lib/connection/drainingreactivesocket.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    reactivesocket._pendings = 0;
+    reactivesocket._closedCb = null;
+
+    var underlyingRequest = reactivesocket.request.bind(reactivesocket);
```
Comment (yunong) [inline lib/connection/drainingreactivesocket.js]: I know you're overriding functions from the base reactivesocket object. Have you looked into ES6 classes? that may make this much easier. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes

File: lib/connection/reactivesocketfactory.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+'use strict';
```
Comment (yunong) [inline lib/connection/reactivesocketfactory.js]: I think you want the name of this file to be reactiveSocketFactory.js in keeping convention with the rest of this repo

File: lib/connection/reactivesocketfactory.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+var DEFAULT_OPTIONS = {
+    transport: {
+        framed: true
+    },
+    type: 'client',
+    metadataEncoding: 'utf-8',
+    dataEncoding: 'utf-8'
+};
+
+function ReactiveSocketFactory(opts) {
```
Comment (yunong) [inline lib/connection/reactivesocketfactory.js]: jsdocs please

File: lib/connection/reactivesocketfactory.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    EventEmitter.call(this);
+}
+
+util.inherits(ReactiveSocketFactory, EventEmitter);
+
+module.exports = ReactiveSocketFactory;
+
+/// API
+
+ReactiveSocketFactory.prototype.apply = function apply(req) {
```
Comment (yunong) [inline lib/connection/reactivesocketfactory.js]: Did you mean to name this function `apply`? This overrides the existing `apply` method on all JS objects: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply

File: lib/connection/weightedSocket.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+var STARTUP_PENALTY = Math.floor(Number.MAX_SAFE_INTEGER / 2);
+
+var CLOCK = {
+    now: function () {
+        var t = process.hrtime();
+        return (1e9 * t[0] + t[1]) / (1000 * 1000);
+    }
+};
+
+function WeightedSocket(reactivesocket, opts) {
```
Comment (yunong) [inline lib/connection/weightedSocket.js]: JS Docs please

File: lib/common/frugalMedian.js (hunk 1 above)
Comment (stevegury) [inline lib/common/frugalMedian.js]: Done

File: lib/connection/drainingreactivesocket.js (hunk 2 above)
Comment (stevegury) [inline lib/connection/drainingreactivesocket.js]: Done

File: lib/connection/reactivesocketfactory.js (hunk 4 above)
Comment (stevegury) [inline lib/connection/reactivesocketfactory.js]: Done

File: lib/connection/reactivesocketfactory.js (hunk 5 above)
Comment (stevegury) [inline lib/connection/reactivesocketfactory.js]: Done

File: lib/connection/reactivesocketfactory.js (hunk 6 above)
Comment (stevegury) [inline lib/connection/reactivesocketfactory.js]: I didn't know about that, I renamed it to `build`

File: lib/connection/weightedSocket.js (hunk 7 above)
Comment (stevegury) [inline lib/connection/weightedSocket.js]: Done

File: lib/connection/weightedSocket.js [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+ * @param {Object} opts.inactivityPeriodMs period in ms, after which the
+ *      algorithm will start decaying the value.
+ * @param {Number} opts.medianBufferSize Number of elements to keep in the
+ *      buffer when estimating the median
+ *
+ * This function override the function from the underlying ReactiveSocket.
+ *
+ * @returns {WeightedSocket}
+ */
+
```
Comment (yunong) [inline lib/connection/weightedSocket.js]: extra new line

**Pull request merged**